cl/audio/tasks.py

```diff
@@ -45,3 +51,55 @@ def upload_audio_to_ia(self, af_pk: int) -> None:
         af.save()
     else:
         increment_failure_count(af)
+
+
+@app.task
+def transcribe_from_open_ai_api(audio_pk: int):
+    """Get transcription from OpenAI API whisper-1 model
+    openai.OpenAI() client expects the environment variable OPENAI_API_KEY to be set
+
+    If successful, updates Audio object and creates related AudioTranscriptionMetadata object
+
+    :param audio_pk: audio object primary key
+    """
+    audio = Audio.objects.get(pk=audio_pk)
+    audio_file = audio.local_path_mp3
+    file = (audio_file.name.split("/")[-1], audio_file.read(), "mp3")
+
+    logger.info("Starting transcription for audio %s", audio_pk)
+
+    with openai.OpenAI() as client:
+        try:
+            transcript = client.audio.transcriptions.create(
+                file=file,
+                model="whisper-1",
+                language="en",
+                response_format="verbose_json",
+                timestamp_granularities=["word", "segment"],
+                prompt=audio.case_name,
+            )
+        except openai.RateLimitError:
+            # implement backoff strategy
+            # More about openai errors
+            # https://github.com/openai/openai-python?tab=readme-ov-file#handling-errors
+            pass
```

Comment:
Celery makes catching/retrying pretty easy. Look around in the code for a bunch of examples.

Reply:
I found that we have 2 options to handle API errors (including the rate limiting). 
- using  `openai.OpenAI` client `max_retries`
- using celery

Perhaps you prefer one of them

With celery. Note that in this example I am missing some OpenAI API error types, but the general idea is there:
```
@app.task(bind=True, max_retries=5)
def transcribe_from_open_ai_api(audio_pk: int):
    with openai.OpenAI(max_retries=0) as client:
        try:
            transcript = client.audio.transcriptions.create(
                file=file,
                model="whisper-1",
                language="en",
                response_format="verbose_json",
                timestamp_granularities=["word", "segment"],
                prompt=audio.case_name,
            )
        # sleep at least one minute before retrying
        except openai.RateLimitError:
            self.retry(countdown=60*(1+self.request.retries))
        except (openai.APIConnectionError):
            self.retry(countdown=60)
        except openai.UnprocessableEntityError:
            audio.stt_status = Audio.STT_FAILED
            audio.save()
            logger.warning("UnprocessableEntityError for audio %s", audio_pk)
            return
            
         ....
```
------
With the python `openai` package. They handle [retries](https://github.com/openai/openai-python?tab=readme-ov-file#retries) for "Connection errors, 408 Request Timeout, 409 Conflict, 429 Rate Limit, and>=500 Internal errors". It also checks the explicit header `x-should-retry` from the server : 
https://github.com/openai/openai-python/blob/54a5911f5215148a0bdeb10e2bcfb84f635a75b9/src/openai/_base_client.py#L679-L712

```
@app.task
def transcribe_from_open_ai_api(audio_pk: int):
    audio = Audio.objects.get(pk=audio_pk)
    audio_file = audio.local_path_mp3
    file = (audio_file.name.split("/")[-1], audio_file.read(), "mp3")

    logger.info("Starting transcription for audio %s", audio_pk)

    with openai.OpenAI(max_retries=5) as client:
        try:
            transcript = client.audio.transcriptions.create(
                file=file,
                model="whisper-1",
                language="en",
                response_format="verbose_json",
                timestamp_granularities=["word", "segment"],
                prompt=audio.case_name,
            )
        except openai.UnprocessableEntityError:
            audio.stt_status = Audio.STT_FAILED
            audio.save()
            logger.warning("UnprocessableEntityError for audio %s", audio_pk)
            return
        except openai.OpenAIError as e:
            # Sending to Sentry errors that are not auto-retried
            # or that have failed many times
            capture_exception(e)
            return
```

The problem with the packaged retry is that at most it waits for 60 seconds, which is the lower limit of waiting that I expect us to need.

Reply:
Definitely better to use the celery retry. You don't want your celery workers sleeping because that would stop other work from proceeding.